ipam/calico-ipam.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/projectcalico/libcalico/lib/ipam"
+)
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdDel)
+}
+
+// IPAMConfig represents the IP related network configuration.
+type IPAMConfig struct {
+	Name          string
+	Type          string `json:"type"`
+	EtcdAuthority string `json:"etcd_authority"`
+	EtcdEndpoints string `json:"etcd_endpoints"`
+	AssignIpv4    *bool  `json:"assign_ipv4"` // TODO make sure this defaults to true
+	AssignIpv6    *bool  `json:"assign_ipv6"`
+
+	Args *ipamArgs `json:"-"`
+}
+
+type ipamArgs struct {
+	types.CommonArgs
+	IP net.IP `json:"ip,omitempty"`
+}
+
+type netConf struct {
+	Name string      `json:"name"`
+	IPAM *IPAMConfig `json:"ipam"`
+}
+
+func loadIPAMConfig(bytes []byte, args string) (*IPAMConfig, error) {
+	n := netConf{}
+	if err := json.Unmarshal(bytes, &n); err != nil {
+		return nil, err
+	}
+
+	if args != "" {
+		n.IPAM.Args = &ipamArgs{}
+		err := types.LoadArgs(args, n.IPAM.Args)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if n.IPAM == nil {
+		return nil, fmt.Errorf("missing 'ipam' key")
+	}
+
+	// Copy net name into IPAM so not to drag Net struct around
+	n.IPAM.Name = n.Name
+
+	return n.IPAM, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := loadIPAMConfig(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	ipamClient, err := ipam.NewIPAMClient()
+	if err != nil {
+		return err
+	}
+
+	// Default to assigning an IPv4 address
+	num4 := 1
+	if conf.AssignIpv4 != nil && *conf.AssignIpv4 == false {
+		num4 = 0
+	}
+
+	// Default to NOT assigning an IPv6 address
+	num6 := 0
+	if conf.AssignIpv6 != nil && *conf.AssignIpv6 == true {
+		num6 = 1
+	}
+
+	// TODO - Read the IP from CNI_ARGS and use it.
+	// TODO - Use the workloadID as the handle (i.e. need to know about k8s)
+	// TODO - plumb through hostname
+	// TODO - plumb through etcd auth
+	// TODO - confirm with Casey if HandleID really needs to be a pointer to a string.
```

Comment:
The Python code certainly differentiated between "None" and empty string, so I make the Go code understand that distinction as well.  

I think this will be needed for the upgrade case, but I haven't tested it yet.  Let's still chat about this when I'm in the office.